lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,770 @@
+    size_t lb = op.getLowBit();
+
+    // Generate the slice instruction
+    genSlice(op, op0, lb, w);
+  }
+
+  // Btor2 uses similar syntax as hw for its comparisons
+  // So we simply need to emit the cmpop name and check for corner cases
+  // where the namings differ.
+  void visitComb(comb::ICmpOp op) {
+    Value lhs = op.getOperand(0);
+    Value rhs = op.getOperand(1);
+
+    // Extract the predicate name (assuming that its a valid btor2
+    // predicate)
+    StringRef pred = stringifyICmpPredicate(op.getPredicate());
+
+    // Check for special cases where hw doesn't align with btor syntax
+    if (pred == hwNeqStr)
+      pred = neqStr;
+
+    // Width of result is always 1 for comparison
+    genSort(bitvecStr, 1);
+
+    // With the special cases out of the way, the emission is the same as that
+    // of a binary op
+    genBinOp(pred, op, lhs, rhs, 1);
+  }
+
+  // Muxes generally convert to an ite statement
+  void visitComb(comb::MuxOp op) {
+    // Extract predicate, true and false values
+    Value pred = op.getCond();
+    Value tval = op.getTrueValue();
+    Value fval = op.getFalseValue();
+
+    // We assume that both tval and fval have the same width
+    // This width should be the same as the output width
+    int64_t w = requireSort(op.getType());
+
+    // Generate the ite instruction
+    genIte(op, pred, tval, fval, w);
+  }
+
+  void visitComb(Operation *op) { visitInvalidComb(op); }
+
+  void visitInvalidComb(Operation *op) {
+    // try sv ops
+    dispatchSVVisitor(op);
+  }
+
+  // Assertions are negated then converted to a btor2 bad instruction
+  void visitSV(sv::AssertOp op) {
+    // Expression is what we will try to invert for our assertion
+    Value expr = op.getExpression();
+
+    // This sort is for assertion inversion and potential implies
+    genSort(bitvecStr, 1);
+
+    // Check for an overaching enable
+    // In our case the sv.if operation will probably only be used when
+    // conditioning an sv.assert on an enable signal. This means that
+    // its condition is probably used to imply our assertion
+    if (auto ifop = dyn_cast<sv::IfOp>(((Operation *)op)->getParentOp())) {
+      Value en = ifop.getOperand();
+
+      // Generate the implication
+      genImplies(ifop, en, expr);
+
+      // Generate the implies inversion
+      genUnaryOp(op, ifop, notStr, 1);
+    } else {
+      // Generate the expression inversion
+      genUnaryOp(op, expr, notStr, 1);
+    }
+
+    // Genrate the bad btor2 intruction
+    genBad(op);
+  }
+  // Assumptions are converted to a btor2 constraint instruction
+  void visitSV(sv::AssumeOp op) {
+    // Extract the expression that we want our constraint to be about
+    Value expr = op.getExpression();
+    genConstraint(expr);
+  }
+
+  void visitSV(Operation *op) { visitInvalidSV(op); }
+
+  void visitInvalidSV(Operation *op) {
+    // The only op left is registers
+    auto reg = dyn_cast<seq::FirRegOp>(op);
+    if (!reg)
+      return;
```

Comment:
I'm not sure what the expectation of this pass is, but should it throw an error on unsupported operations? Or is silently ignoring them a sensible behavior?

Reply:
The idea is that some operations are explicitly ignored in btor2 (like outputs, wires, sv.if, sv.macro, sv.always, etc...), so I wanted to be able to ignore them without having to write out a bunch of empty visitors. Maybe I can throw an a exception for unsupported operations and manually add visitors for explicitly ignored ops like outputs, that would work as well, so I'll do that if it's preferred.

Reply:
I don't think you need to explicitly do nothing for all the ops you don't care about--you are free to only define visitors that matter. I just wasn't sure if the output would make sense to skip over such ops, but if that's how btor2 works, that's great. Sorry for giving you the run around--I just looked at the most recent commit, and I don't think you need to do that.

Reply:
Well there are some operations that I plan to support in the future because ignoring them wouldn't produce a correct behavior, so I think that differentiating between ops that I can ignore and those that I simply haven't implemented yet is a sound thing to do, I was just afraid that it would make the code ugly so I didn't do it at first. So I think that having this explicit difference between the two is better, but I can revert it if preferred, just let me know.

Reply:
Gotcha, well I will leave it up to you.

What I would do is revert to what you had before--having big TypeSwitches for what's supported kind of defeats some of the niceness of the Visitors. And in the final case where this comment is rooted, throw an error that says something like "unsupported op". This will throw if one of the ops you plan to support but haven't finished yet is encountered, but that's fine--when you add support, you just define the visitMyOp method. It will also throw for ops you never want to see, but I think that's also fine. This converter already expects the IR to be in a certain shape (i.e., modules fully flattened), and expecting those ops to not be in the IR also seems reasonable. This way the converter just defines visitors for what it's supposed to handle. To start, that is limited, but it can grow over time without large allow/deny lists of ops.

Reply:
What about the way I have it now in  `visitUnsupportedOp`(line 748). It has a big typeswitch, but only for what I need to ignore, if I throw an error there the conversion simply won't work. So the alternative would be to have a bunch of empty visitors for each operation that has to be ignore and should be allowed in the input (like `hw.module` or `hw.output` or `sv.always`, or many others) which seems unnecessary to me and not as concise as a typeswitch. But again, I will write out the empty visitors if this is usually what's expected from this type of pass.

Reply:
It is up to you, what you have now seems ok modulo some comments I just left. I still think having an allowlist of ops that are ok to ignore can be fragile. I'm not really sure what your pass expects to handle, but could you do a more generic check, like "if comb dialect, throw an error, else ignore"? That might be a slightly nicer flavor of the current setup.

> if I throw an error there the conversion simply won't work

I don't really follow this, so let me try to explain what I meant before. You have a brand new pass, and it's totally normal for it to not support everything you plan to support. It's also totally normal for it to only work on certain IR. But you also have control over what input you feed it and what you expect it to work on.

For example, you assume the module hierarchy is flattened, so instead of calling walk on the `hw.module` (which also visits the `hw.module` op itself), you can loop over `hwModule->getBodyBody()->without_terminator()`. That way your visitor doesn't have to deal with `hw.module` or `hw.output` at all. You have some control here, and that can help.

For another example, why is `sv.always` allowed in the input if it doesn't make sense for this pass? Couldn't you assume whatever flattened the module hierarchy scrubbed `sv.always` from the input as well? Nothing is calling your pass right now besides the unit tests you're writing, so why not just avoid using `sv.always` in those tests?

This is how I'd set up a brand new conversion. Visitors for what is supported, and errors if preconditions are not met, or unsupported ops are visited. To start, you don't have to support everything. And you don't have to explicitly list every op you don't care about--you can assume for now that they aren't in the IR, since you're writing the unit tests. Eventually, for this to be part of a larger pipeline/tool, something will have to hammer the IR into the shape this expects, but you can just assume that exists to get started. That way, you can just start with visitors for what you want to start with, and unit tests that exercise those.

Maybe I'm missing something. If you're trying to run this pass on real IR that you're getting from some pipeline, and you want it to dump out what it can but not error on unsupported things, that's a different story. In that case I'd just have visitUnhandledOp do nothing, and not worry about differentiating between ops you know you don't care about versus ops you haven't implemented. You could also do something like print a warning in debug mode.

I just have a slightly allergic reaction to an allowlist of every op you know you can safely ignore, either in a TypeSwitch or in visit methods that do nothing. In summary, it's up to you, but I'd recommend either having visitUnhandledOp always throw an error or always do nothing (and maybe print a warning).

Reply:
Maybe `sv.always` isn't a good example, since it looks like you actually do expect that in the IR. In that case, I think a visitor that does nothing is appropriate. I'm more thinking about random ops you don't ever want to visit, hopefully that was clear. Like `sv.some_wonky_vivado_thing`.

Reply:
This conversation got me thinking, so I [looked up](https://github.com/search?q=repo%3Allvm%2Fcirct%20visitUnhandled&type=code) what other CIRCT visitors do.

The LEC tool will throw errors: https://github.com/llvm/circt/blob/04260957c69b1ac91c46bdcd690aa670038bbce8/lib/LogicalEquivalence/LogicExporter.cpp#L64

ExportVerilog with emit an error message: https://github.com/llvm/circt/blob/04260957c69b1ac91c46bdcd690aa670038bbce8/lib/Conversion/ExportVerilog/ExportVerilog.cpp#L3304

But It's also in the middle of writing to an output stream, so it will write something invalid and then move on, which is a little funky.

This pass does generic check for all the ops it doesn't explicitly handle, and errors if they don't meet the pass' post condition: https://github.com/llvm/circt/blob/04260957c69b1ac91c46bdcd690aa670038bbce8/lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp#L174